File: UbuntuLauncher/app/src/main/java/be/robinj/ubuntu/unity/dash/lens/GitHub.java
package be.robinj.ubuntu.unity.dash.lens;

import android.content.Context;
import android.graphics.drawable.Drawable;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.IOException;
import java.net.URLEncoder;
import java.util.ArrayList;
import java.util.List;

import be.robinj.ubuntu.R;

/**
 * Created by robin on 4/11/14.
 */
public class GitHub extends Lens
{
	private final String API = "https://api.github.com/search/repositories?q={:QUERY:}";

	public GitHub (Context context)
	{
		super (context);

		this.icon = context.getResources ().getDrawable (R.drawable.dash_search_lens_github);
	}

	public String getName ()
	{
		return "GitHub";
	}

	public String getDescription ()
	{
		return "GitHub repository search results";
	}

	public List<LensSearchResult> search (String str) throws IOException, JSONException
	{
		String apiResults = this.downloadStr (this.API.replace ("{:QUERY:}", URLEncoder.encode (str, "UTF-8")));

		JSONObject json = new JSONObject (apiResults);
		JSONArray items = json.getJSONArray ("items");
		List<LensSearchResult> results = new ArrayList<LensSearchResult> ();

		for (int i = 0; i < items.length (); i++)
		{
			JSONObject item = items.getJSONObject (i);

			if (item.has ("full_name") && item.has ("html_url"))
			{
				LensSearchResult result = new LensSearchResult (this.context, item.getString ("full_name"), item.getString ("html_url"), this.icon);

				results.add (result);
			}
		}

		return results;
	}
}

File: app/src/main/java/be/robinj/distrohopper/desktop/dash/lens/GitHub.kt
package be.robinj.distrohopper.desktop.dash.lens

import android.content.Context
import be.robinj.distrohopper.R
import org.json.JSONObject
import java.net.URLEncoder

/**
 * Created by robin on 4/11/14.
 */
open class GitHub(context: Context) : Lens(context) {

    init {
        this.icon = context.resources.getDrawable(R.drawable.dash_search_lens_github)
    }

    override val type = LensType.NETWORK

    override fun getName() = "GitHub"

    override fun getDescription() = "GitHub repository search results"

    override suspend fun search(query: String, maxResults: Int, emitter: LensResultEmitter) {
        val apiResults = this.downloadStr(API.replace("{:QUERY:}", URLEncoder.encode(query, "UTF-8")))

        val items = JSONObject(apiResults).getJSONArray("items")
        var nResults = 0

        for (i in 0 until items.length()) {
            val item = items.getJSONObject(i)

            if (item.has("full_name") && item.has("html_url")) {
                emitter.emit(LensSearchResult(
                    this.context, item.getString("full_name"), item.getString("html_url"), this.icon))

                if (++nResults >= maxResults) {
                    break
                }
            }
        }
    }

    override fun onClick(url: String) = this.openInBrowser(url)

    companion object {
        private const val API = "https://api.github.com/search/repositories?q={:QUERY:}"
    }
}
